web3-functions/pyth/index.ts

```diff
@@ -0,0 +1,315 @@
+
+function shouldFetchPythConfig(pythConfigStorage: any) {
+  const isNotFoundInStorage = pythConfigStorage.pythConfig === void 0;
+  return isNotFoundInStorage || Date.now() / 1e3 - pythConfigStorage.timestamp > pythConfigStorage.pythConfig.configRefreshRateInSeconds;
+}
+
+async function fetchPythConfigIfNecessary(storage: any, gistId: string) {
+  const octokit = new Octokit();
+  let pythConfig;
+  let pythConfigStorage = JSON.parse(
+    await storage.get("pythConfig") ?? "{}"
+  );
+  if (shouldFetchPythConfig(pythConfigStorage)) {
+    const gistDetails = await octokit.rest.gists.get({ gist_id: gistId });
+    const files = gistDetails.data.files;
+    if (!files)
+      throw new Error(`No files in gist`);
+    for (const file of Object.values(files)) {
+      if (file?.filename === "config.yaml" && file.content) {
+        pythConfig = YAML.parse(file.content);
+        break;
+      }
+    }
+    if (!pythConfig)
+      throw new Error(`No config.yaml loaded for PythConfig`);
+    pythConfigStorage = {
+      timestamp: Date.now() / 1e3,
+      pythConfig
+    };
+    const pythConfigStorageValue = JSON.stringify(pythConfigStorage);
+    if (pythConfig.debug) {
+      console.debug(
+        `storing fetched pythConfigStorageValue: ${pythConfigStorageValue}`
+      );
+    }
+    await storage.set("pythConfig", pythConfigStorageValue);
+  } else {
+    pythConfig = pythConfigStorage.pythConfig;
+    if (pythConfig.debug) {
+      console.debug("using pythConfig from storage");
+    }
+  }
+  return pythConfig;
+}
+async function getCurrentPrices(priceIds: string[], connection: any, debug: boolean) {
+  const latestPriceFeeds = await connection.getLatestPriceFeeds(priceIds);
+  if (latestPriceFeeds === void 0) {
+    return void 0;
+  }
+  if (debug) {
+    console.debug(`latestPriceFeeds: ${JSON.stringify(latestPriceFeeds)}`);
+  }
+  return latestPriceFeeds.map((pf: any) => {
+    return {
+      id: pf.id,
+      price: pf.getPriceUnchecked(),
+    };
+  }).filter((pf: any) => {
+    return pf !== void 0 && pf.price !== void 0;
+  }).reduce((acc: any, pf: any) => {
+    acc.set(addLeading0x(pf.id), pf.price);
+    return acc;
+  }, /* @__PURE__ */ new Map());
+}
+
+async function getLastPricesFromChain(provider: any, oracleAddress: string, priceIds: string[], ) {
+  let result = new Map();
+  let oracle;
+  for (const priceId of priceIds) {
+    oracle = new Contract(oracleAddress, ORACLE_ABI, provider);
+    const priceInfo = await oracle.getPriceUnsafe(priceId);
+    // decode result
+    const price = priceInfo[0].toString();
+    const expo = parseInt(priceInfo[2].toString());
+    const publishTime = parseInt(priceInfo[3].toString());
+    const priceInfoJson = {
+      price,
+      expo,
+      publishTime
+    }
+    result.set(priceId, priceInfoJson);
+  }
+  return result;
+}
+
+function generatePriceUpdateList(
+  priceIds: any,
+  currentPrices: any,
+  lastPrices: any,
+  allPriceIdsByItem: { [key: string]: string[] },
```

Comment:
I am not sure I understand the full purpose of this one

Reply:
Actually I'd like to keep this one as it is useful when fetching the prices :
 const allPriceIds = [...new Set(Object.values(allPriceIdsByItem).flat())];

Reply:
and I use allPriceIds a lot